File: frontend/src/api/rating/rating.ts
import { get, post } from '../request'

import type {
  AddRatingItemRequest,
  PageResult,
  QueryRatingResultParams,
  RatingItem,
  RatingItemQueryParams,
  RatingResultItem,
  RatingStatistics,
  UpdateRatingItemRequest,
} from '@/types'

/**
 * 获取评分项目列表。
 */
export function getRatingItemList(params: RatingItemQueryParams) {
  return get<PageResult<RatingItem>>('/rating/items', params as unknown as Record<string, unknown>)
}

/**
 * 新增评分项目。
 *
 * 新增时必须指定所属 Topic。
 */
export function createRatingItem(data: AddRatingItemRequest) {
  return post<RatingItem>('/rating/addItem', data)
}

/**
 * 修改评分项目。
 *
 * RatingItem 创建后不允许修改所属 Topic。
 */
export function updateRatingItem(data: UpdateRatingItemRequest) {
  return post<RatingItem>('/rating/updateItem', data)
}

/**
 * 删除评分项目。
 */
export function deleteRatingItem(id: number) {
  return post<void>('/rating/deleteItem', {
    id,
  })
}

/**
 * 获取单个评分项目。
 */
export function getRatingItem(params: { id: number }) {
  return get<RatingItem>('/rating/getItem', params)
}

/**
 * 开始评分。
 */
export function startRating(data: { id: number }) {
  return post<RatingItem>('/rating/startRating', data)
}

/**
 * 结束评分。
 */
export function finishRating(data: { id: number }) {
  return post<RatingItem>('/rating/finishRating', data)
}

/**
 * 获取实时评分统计。
 */
export function getRatingStatistics(params: { id: number }) {
  return get<RatingStatistics>('/rating/getStatistics', params)
}

/**
 * 当前客户端针对某个 RatingItem
 * 的评分提交状态。
 */
export interface RatingClientStatus {
  submitted: boolean

  /**
   * 专家：
   * 0 ~ 100
   *
   * 大众：
   * 1 或 2，表示点赞数量。
   *
   * 未评分：
   * null
   */
  score: number | null

  submitTime: string | null
}

export interface GetRatingStatusParams {
  ratingItemId: number
  clientId: string
}

/**
 * 提交评分请求。
 *
 * score：
 *
 * 专家评委：
 * 0 ~ 100
 *
 * 大众评委：
 * 1 或 2
 */
export interface SubmitScoreRequest {
  ratingItemId: number

  clientId: string

  score: number

  /**
   * 大众评委不传。
   *
   * 专家评委从 Topic 专家二维码
   * URL 中获取。
   */
  expertToken?: string
}

/**
 * 查询当前浏览器客户端是否已经评分。
 */
export function getRatingStatus(params: GetRatingStatusParams) {
  return get<RatingClientStatus>('/rating/getRatingStatus', params)
}

/**
 * 提交评分。
 */
export function submitScore(data: SubmitScoreRequest) {
  return post('/rating/submitScore', data)
}

/**
 * 查询评分结果列表。
 */
export function queryRatingResults(params: QueryRatingResultParams): Promise<PageResult<RatingResultItem>> {
  return get<PageResult<RatingResultItem>>('/rating/queryResults', params)
}

/**
 * RatingItem 图片上传结果。
 */
export interface UploadItemImageResponse {
  url: string
}

/**
 * 上传 RatingItem 图片。
 *
 * 文件上传使用 multipart/form-data，
 * 不能使用普通 JSON post() 封装。
 */
/**
 * 上传 RatingItem 图片。
 */
export interface UploadItemImageResponse {
  url: string
}

/**
 * 上传 RatingItem 图片。
 */
export function uploadItemImage(file: File): Promise<UploadItemImageResponse> {
  const formData = new FormData()

  formData.append('file', file, file.name)

  return post<UploadItemImageResponse>('/rating/uploadItemImage', formData)
}

File: backend/app/api/rating.py
from typing import Annotated

from fastapi import APIRouter, Query, status

from app.database.session import SessionDep
from app.schemas.common import ApiResponse
from app.schemas.rating import (
    CreateRatingItemRequest,
    RatingItemResponse,
    RatingStatus,
    UpdateRatingItemRequest,
    DeleteRatingItemRequest,
    RatingActionRequest,
    RatingStatisticsResponse,
    RatingResultResponse,
    SubmitScoreRequest,
    RatingStatusResponse, RatingResultListItemResponse
)
from app.services.rating_service import RatingService
from app.schemas.common import PageResult

router = APIRouter(
    prefix="/rating",
    tags=["Rating"],
)


@router.get(
    "/items",
    response_model=ApiResponse[
        PageResult[RatingItemResponse]
    ],
)
def get_rating_item_list(
        db: SessionDep,

        topic_id: Annotated[
            int | None,
            Query(
                alias="topicId",
                ge=1,
                description="评分主题 ID",
            ),
        ] = None,

        name: Annotated[
            str | None,
            Query(
                max_length=50,
                description="项目名称",
            ),
        ] = None,

        status_: Annotated[
            RatingStatus | None,
            Query(
                alias="status",
                description="项目状态",
            ),
        ] = None,

        page: Annotated[
            int,
            Query(
                ge=1,
                description="页码",
            ),
        ] = 1,

        page_size: Annotated[
            int,
            Query(
                alias="pageSize",
                ge=1,
                le=100,
                description="每页数量",
            ),
        ] = 10,
) -> ApiResponse[
    PageResult[RatingItemResponse]
]:
    """
    分页查询评分项目。
    """

    service = RatingService(db)

    result = service.list_items(
        topic_id=topic_id,
        name=name,
        status=status_,
        page=page,
        page_size=page_size,
    )

    return ApiResponse(
        data=result,
    )


@router.post(
    "/addItem",
    response_model=ApiResponse[
        RatingItemResponse
    ],
    status_code=status.HTTP_201_CREATED,
)
def create_rating_item(
        request: CreateRatingItemRequest,
        db: SessionDep,
) -> ApiResponse[RatingItemResponse]:
    """
    新增评分项目。
    """

    service = RatingService(db)

    item = service.create_item(
        request
    )

    return ApiResponse(
        message="新增成功",
        data=item,
    )


@router.post(
    "/updateItem",
    response_model=ApiResponse[
        RatingItemResponse
    ],
)
def update_rating_item(
        request: UpdateRatingItemRequest,
        db: SessionDep,
) -> ApiResponse[RatingItemResponse]:
    """
    修改评分项目。
    """

    service = RatingService(db)

    item = service.update_item(
        request=request,
    )

    return ApiResponse(
        message="修改成功",
        data=item,
    )


@router.post(
    "/deleteItem",
    response_model=ApiResponse[bool],
)
def delete_rating_item(
        request: DeleteRatingItemRequest,
        db: SessionDep,
) -> ApiResponse[bool]:
    """
    删除评分项目。
    """

    service = RatingService(db)

    service.delete_item(
        request=request,
    )

    return ApiResponse(
        message="删除成功",
        data=True,
    )


@router.get(
    "/getItem",
    response_model=ApiResponse[
        RatingItemResponse
    ],
)
def get_rating_item(
        id: Annotated[
            int,
            Query(
                ge=1,
                description="评分项目 ID",
            ),
        ],
        db: SessionDep,
) -> ApiResponse[RatingItemResponse]:
    """
    查询单个评分项目。
    """

    service = RatingService(db)

    item = service.get_item(
        item_id=id,
    )

    return ApiResponse(
        data=item,
    )


@router.post(
    "/startRating",
    response_model=ApiResponse[
        RatingItemResponse
    ],
)
def start_rating(
        request: RatingActionRequest,
        db: SessionDep,
) -> ApiResponse[RatingItemResponse]:
    """
    开始评分。
    """

    service = RatingService(db)

    item = service.start_rating(
        item_id=request.id,
    )

    return ApiResponse(
        message="评分已开始",
        data=item,
    )


@router.post(
    "/finishRating",
    response_model=ApiResponse[
        RatingItemResponse
    ],
)
def finish_rating(
        request: RatingActionRequest,
        db: SessionDep,
) -> ApiResponse[RatingItemResponse]:
    """
    结束评分。
    """

    service = RatingService(db)

    item = service.finish_rating(
        item_id=request.id,
    )

    return ApiResponse(
        message="评分已结束",
        data=item,
    )


@router.get(
    "/getStatistics",
    response_model=ApiResponse[
        RatingStatisticsResponse
    ],
)
def get_statistics(
        db: SessionDep,
        id: int = Query(
            ge=1,
            description="评分项目 ID",
        ),
) -> ApiResponse[
    RatingStatisticsResponse
]:
    """
    获取评分项目实时统计结果。
    """

    service = RatingService(db)

    result = service.get_statistics(
        item_id=id,
    )

    return ApiResponse(
        data=result,
    )


@router.post(
    "/submitScore",
    response_model=ApiResponse[
        RatingResultResponse
    ],
)
def submit_score(
        request: SubmitScoreRequest,
        db: SessionDep,
) -> ApiResponse[RatingResultResponse]:
    """
    提交评分。
    """

    service = RatingService(db)

    result = service.submit_score(
        request=request,
    )

    return ApiResponse(
        message="评分提交成功",
        data=result,
    )


@router.get(
    "/getRatingStatus",
    response_model=ApiResponse[
        RatingStatusResponse
    ],
)
def get_rating_status(
        db: SessionDep,
        ratingItemId: int = Query(
            ge=1,
            description="评分项目 ID",
        ),
        clientId: str = Query(
            min_length=1,
            max_length=64,
            description="浏览器客户端 ID",
        ),
) -> ApiResponse[RatingStatusResponse]:
    """
    查询当前浏览器客户端的评分状态。
    """

    service = RatingService(db)

    result = service.get_rating_status(
        item_id=ratingItemId,
        client_id=clientId,
    )

    return ApiResponse(
        data=result,
    )


@router.get(
    "/queryResults",
    response_model=ApiResponse[
        PageResult[
            RatingResultListItemResponse
        ]
    ],
)
def query_results(
        db: SessionDep,
        page: int = Query(
            default=1,
            ge=1,
        ),
        pageSize: int = Query(
            default=10,
            ge=1,
            le=100,
        ),
        topicId: int | None = Query(
            default=None,
        ),
        itemName: str | None = Query(
            default=None,
        ),
        reviewerType: int | None = Query(
            default=None,
            ge=0,
            le=1,
        ),
        score: float | None = Query(
            default=None,
            ge=0,
            le=100,
        ),
):
    """
    分页查询评分结果。
    """

    service = RatingService(db)

    result = service.query_results(
        page=page,
        page_size=pageSize,
        topic_id=topicId,
        item_name=itemName,
        reviewer_type=reviewerType,
        score=score,
    )

    return ApiResponse(
        data=result,
    )
